qualtran/bloqs/arithmetic/multiplication.py

```diff
@@ -12,16 +12,80 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
-from typing import Set, TYPE_CHECKING
+from typing import Any, Dict, Iterable, Sequence, Set, TYPE_CHECKING, Union
 
+import cirq
 from attrs import frozen
 
-from qualtran import Bloq, Register, Side, Signature
+from qualtran import Bloq, GateWithRegisters, Register, Side, Signature
 from qualtran.bloqs.basic_gates import Toffoli
 from qualtran.cirq_interop.t_complexity_protocol import TComplexity
 
 if TYPE_CHECKING:
+    from qualtran import SoquetT
     from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+    from qualtran.simulation.classical_sim import ClassicalValT
+
+
+@frozen
+class PlusEqualProduct(GateWithRegisters, cirq.ArithmeticGate):
+    """Performs result += a * b"""
+
+    a_bitsize: int
+    b_bitsize: int
+    result_bitsize: int
+    adjoint: bool = False
+
+    def short_name(self) -> str:
+        return "result -= a*b" if self.adjoint else "result += a*b"
+
+    @property
+    def signature(self) -> 'Signature':
+        return Signature.build(a=self.a_bitsize, b=self.b_bitsize, result=self.result_bitsize)
+
+    def registers(self) -> Sequence[Union[int, Sequence[int]]]:
+        return [2] * self.a_bitsize, [2] * self.b_bitsize, [2] * self.result_bitsize
+
+    def apply(self, a: int, b: int, result: int) -> Union[int, Iterable[int]]:
+        return a, b, (result + a * b * ((-1) ** self.adjoint)) % (2**self.result_bitsize)
+
+    def with_registers(self, *new_registers: Union[int, Sequence[int]]):
+        raise NotImplementedError("Not needed.")
```

Comment:
?

Reply:
`with_registers` is an abstract method on `cirq.ArithmeticGate` which isn't really necessary to be implemented. It's present due to legacy reasons. I think we should have a `ArithmeticGate` base class for Qualtran that derives from `GateWithRegisters` and `cirq.ArithmeticGate` and provides all the shims so that user simply needs to implement `on_classical_vals` as we'd expect from a Qualtran native interface. 

I'll do this as a follow up PR